### PR TITLE
[ETHEREUM-CONTRACTS/SDK-CORE/SUBGRAPH] Use more Typechain capabilities

### PR DESCRIPTION
**Pull request by 0xdavinchee**
* Provide end-consumer access to typechain files via `ethereum-contracts` as the origin source and also exported via `sdk-core`
* no more abi file imports, using typechain generated factories all around
* fix up lint scripts
* add typechain build pipeline in ethereum-contracts

### CODE REVIEW

File: packages/ethereum-contracts/package.json
Comment (surewht) [inline packages/ethereum-contracts/package.json:20]: Will we increase the version in another pr?

Comment (0xdavinchee) [reply packages/ethereum-contracts/package.json:20]: yeah, I think once we want to do a release it will make sense to - once we complete the cfav1 forwarder changes + maybe: https://github.com/superfluid-finance/protocol-monorepo/issues/1112

**Review by surewht — APPROVED** [review]
Awesome^^

**Comment by 0xdavinchee** [comment]
blocked by doc:html consuming the new exported types from ethereum-contracts and polluting the reference docs

File: packages/ethereum-contracts/CHANGELOG.md
Comment (kasparkallas) [inline packages/ethereum-contracts/CHANGELOG.md:10]: What's a "Hook"?

File: packages/sdk-core/package.json
Comment (kasparkallas) [inline packages/sdk-core/package.json:58]: I would think the published dev package will be broken if this is not incremented.

File: packages/sdk-core/src/ERC20Token.ts
Comment (kasparkallas) [inline packages/sdk-core/src/ERC20Token.ts]: Why is it good to connect the contract here? Aren't now all the contracts possibly connected to a signer? And when write operations are invoked then that signer is not used... unless there's an error in our code and we miss passing in the new signer in which case this signer is used.

**Review by kasparkallas — COMMENTED** [review]
1. Please quickly host the TypeDoc somewhere so we could preview it.
2. I wish there was a simple example available for consumers to generate TypeChain themselves from `@superfluid-finance/ethereum-contracts` without the bash script ABI extraction overhead. I could imagine someone wanting to generate TypeChain for hardhat or web3 or something else that is not pre-generated by us.

File: packages/ethereum-contracts/CHANGELOG.md
Comment (0xdavinchee) [reply packages/ethereum-contracts/CHANGELOG.md:10]: check out the #1099 and also https://github.com/superfluid-finance/protocol-monorepo/issues/1096

File: packages/sdk-core/package.json
Comment (0xdavinchee) [reply packages/sdk-core/package.json:58]: > 1. Please quickly host the TypeDoc somewhere so we could preview it.

will do this w/ netlify and post link when done

> 2. I wish there was a simple example available for consumers to generate TypeDoc themselves from `@superfluid-finance/ethereum-contracts` without the bash script ABI extraction overhead. I could imagine someone wanting to generate TypeDoc for hardhat or web3 or something else that is not pre-generated by us.

Are you referring to SDK-CORE here? If yes, it is fairly simple to do this now and the bash script ABI extraction has been deleted from the SDK-Core. I am pretty sure if you have the project locally, you just have to run `yarn build` in ethereum-contracts and then `yarn doc:html` in sdk-core. Then you just have to run `npx serve dist/docs` in sdk-core to have this locally

File: packages/sdk-core/src/ERC20Token.ts
Comment (0xdavinchee) [reply packages/sdk-core/src/ERC20Token.ts]: > Why is it good to connect the contract here? Aren't now all the contracts possibly connected to a signer? 

yes, the contracts may be connected to a signer, but the proper use here is to connect a provider, so we can simply change the allowed type here if you think this is an issue.

> And when write operations are invoked then that signer is not used...

However, let's say we don't do that and we allow the signer and we throw away the signer later, there is no issue with this as you have to do this anyways if you are switching between signers in the same session-one signer at start vs. different signer later at execution is pretty similar to no signer at start then signer on first write operation and different signer on subsequent or each write operation.

> unless there's an error in our code and we miss passing in the new signer in which case this signer is used.

If the user of the product decides to execute transactions via the contract directly, this may be an issue and may warrant only allowing provider type to be passed on initialization. 
However, if they are using our class and creating an `Operation` and doing `.exec()`, this will absolutely fail even if you cast it as any because you would be trying to execute `.sendTransaction` on an undefined object.

File: packages/sdk-core/package.json
Comment (0xdavinchee) [reply packages/sdk-core/package.json:58]: https://visionary-marzipan-1b34ea.netlify.app/

Comment (0xdavinchee) [reply packages/sdk-core/package.json:58]: okay, we can increment the version in this PR

Comment (kasparkallas) [reply packages/sdk-core/package.json:58]: Oops, for 2 I accidentally wrote "TypeDoc" but meant "TypeChain".

Comment (0xdavinchee) [reply packages/sdk-core/package.json:58]: > Oops, for 2 I accidentally wrote "TypeDoc" but meant "TypeChain".

this never required a bash script ABI extraction to do so within `@superfluid-finance/ethereum-contracts`, you could simply run `npx hardhat compile` or `npx hardhat typechain` to get the typechain types for all contracts

File: packages/sdk-core/src/ERC20Token.ts
Comment (kasparkallas) [reply packages/sdk-core/src/ERC20Token.ts]: What's the benefit of connecting here though? It looks like the connection is always overriden anyways.

Comment (0xdavinchee) [reply packages/sdk-core/src/ERC20Token.ts]: Sure that's a fair point.

Firstly: Some extra typecasting is avoided with `.connect`.
Code with .connect:
```
this.contract = ERC20WithTokenInfo__factory.connect(
    this.address,
    providerOrSigner
);
```

Code without `.connect`:
```
this.contract = new ethers.Contract(
    this.address,
    ERC20WithTokenInfo__factory.abi
) as ERC20WithTokenInfo;
```

Secondly, for write operations, I think this will always be overridden and it is probably safe to do this. However, by providing a provider/signer in the initialization of the contract, it opens up the possibility of not requiring the end user to pass a provider on reads. This does become more problematic if the consumer is on multiple chains but this can be pretty easily solved with just passing in a new provider.

Lastly, why not, this operation is not blocking and will not slow anything down - I see no good reason to not use this (imo) cleaner way of initializing a contract.

**Review by kasparkallas — APPROVED** [review]

File: .github/workflows/call.check-query-schema-against-subgraph.yml
Comment (hellwolf) [inline .github/workflows/call.check-query-schema-against-subgraph.yml]: not a blocking comment, but having the same build:* sequence across several places tickles DRY principle. something to tackle later.

**Review by hellwolf — APPROVED** [review]
will unblock this since many people already reviewed.

**Comment by hellwolf** [comment]
> The base branch requires all commits to be signed. [Learn more about signing commits.](https://docs.github.com/articles/about-gpg/)


One could do a signed force commit that squashed everything in one to unblock this.

File: .github/workflows/call.check-query-schema-against-subgraph.yml
Comment (0xdavinchee) [reply .github/workflows/call.check-query-schema-against-subgraph.yml]: will resolve this in the same PR along w/ squashed commit - just does yarn build instead

**Pull request merged**
